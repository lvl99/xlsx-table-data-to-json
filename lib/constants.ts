import { ParsingOptions } from "xlsx";

export const XLSX_DATE_FORMATS_TO_DAYJS_FORMATS: {
  [key: string]: string | string[];
} = {
  "m/d/yy": "M/D/YY",
  "d-mmm-yy": "D-MMM-YY",
  "d-mmm": "D-MMM",
  "mmm-yy": "MMM-YY",
  "h:mm AM/PM": "h:mm A",
  "h:mm:ss AM/PM": "h:mm:ss A",
  "h:mm": "h:mm",
  "h:mm:ss": "h:mm:ss",
  "m/d/yy h:mm": "M/D/YY h:mm",
  "mm:ss": "mm:ss",
  "[h]:mm:ss": ["[h]:mm:ss", "mm:ss"],
  "mmss.0": "mmss.SSS",
  '"上午/下午 "hh"時"mm"分"ss"秒 "': "A hh[時]mm[分]ss[秒]",
  "yyyy-mm-dd": "YYYY-MM-DD",
  "yyyy-mm-dd hh:mm:ss": "YYYY-MM-DD HH:mm:ss",
};

export const DEFAULT_XLSX_PARSING_OPTIONS: ParsingOptions = {
  cellNF: true,
  cellHTML: true,
  cellStyles: true,
};
