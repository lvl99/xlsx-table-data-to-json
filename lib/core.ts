import * as xlsx from "xlsx";
import fetch from "node-fetch";
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";
import get from "lodash/get";
import {
  GenericData,
  TableData,
  TableDataConfig,
} from "table-data-to-json/lib/core";
import convertTableDataToJSON from "table-data-to-json";
import {
  XLSX_DATE_FORMATS_TO_DAYJS_FORMATS,
  DEFAULT_XLSX_PARSING_OPTIONS,
} from "./constants";

export interface XLSXParseConfig {
  url?: string;
  file?: Buffer | string;
  data?: xlsx.WorkBook;
  sheetIndex?: number;
  sheetName?: string;
  parsingOptions?: xlsx.ParsingOptions;
}

export interface XLSXTableDataOutput<Data = GenericData> {
  workbook: xlsx.WorkBook;
  sheet: xlsx.WorkSheet;
  tableData: TableData<Data>;
}

export interface XLSXTableDataConfig
  extends Partial<TableDataConfig>,
    XLSXParseConfig {}

dayjs.extend(utc);

/**
 * Convert XLSX sheet from XLSX workbook to array table data.
 */
export function convertSheetToTableData<Data = GenericData>({
  workbook,
  sheet,
}: {
  workbook: xlsx.WorkBook;
  sheet: xlsx.WorkSheet;
}): TableData<Data> {
  const data: TableData<Data> = [];
  Object.entries(sheet).forEach(([cellRef, cell]) => {
    if (cellRef.indexOf("!") !== 0) {
      const coords = xlsx.utils.decode_cell(cellRef);
      if (!data[coords.r]) data[coords.r] = [];

      let cellValue = cell.v;
      switch (cell.t) {
        case "b":
          cellValue = ["true", "1", "y"].includes(`${cellValue}`.toLowerCase());
          break;

        case "d":
          if (cell.v instanceof Date) {
            cellValue = dayjs.utc(cell.v).toISOString();
          } else {
            cellValue = xlsx.SSF.parse_date_code(cell.v, {
              date1904: workbook.Workbook?.WBProps?.date1904,
            });
          }
          break;

        case "z":
          cellValue = null;
          break;

        case "n":
          if (
            cell.z &&
            Object.keys(XLSX_DATE_FORMATS_TO_DAYJS_FORMATS).includes(cell.z)
          ) {
            const parseDateFormat =
              XLSX_DATE_FORMATS_TO_DAYJS_FORMATS[cell.z] instanceof Array
                ? XLSX_DATE_FORMATS_TO_DAYJS_FORMATS[cell.z][0]
                : (XLSX_DATE_FORMATS_TO_DAYJS_FORMATS[cell.z] as string);

            cellValue = dayjs.utc(cell.w, parseDateFormat).toISOString();
          } else {
            cellValue = parseFloat(cellValue);
          }
          break;

        case "s":
        default:
          cellValue = `${cellValue}`;
      }

      data[coords.r][coords.c] = cellValue;
    }
  });

  return data;
}

export async function convertXLSXToTableData<Data = GenericData>({
  file,
  url,
  data,
  sheetIndex,
  sheetName,
  parsingOptions,
}: XLSXParseConfig): Promise<XLSXTableDataOutput<Data>> {
  const parseConfig = {
    ...DEFAULT_XLSX_PARSING_OPTIONS,
    sheets: sheetIndex || 0,
    ...parsingOptions,
  };

  let workbook: xlsx.WorkBook;
  if (data && data.Sheets) {
    workbook = data;
  } else if (url) {
    // Read remote file
    try {
      const contents = await (await fetch(url)).arrayBuffer();
      workbook = xlsx.read(contents, parseConfig);
    } catch (exp) {
      throw exp;
    }
  } else if (file) {
    // Read local file
    if (file instanceof Buffer) {
      workbook = xlsx.read(file, parseConfig);
    } else {
      workbook = xlsx.readFile(file, parseConfig);
    }
  } else {
    throw Error(
      "Please provide XLSX data, url or file (either file contents buffer or local path string)"
    );
  }

  const sheet: xlsx.Sheet = get(
    workbook.Sheets,
    sheetName || workbook.SheetNames[sheetIndex || 0]
  );
  const tableData: TableData<Data> = convertSheetToTableData<Data>({
    workbook,
    sheet,
  });

  return {
    workbook,
    sheet,
    tableData,
  };
}

/**
 * Convert an XLSX local file path, remote HTTP URL, Buffer,
 * or xlsx.WorkBook object (or other such file format supported
 * by SheetJS) to JSON.
 *
 * Will convert the data to tableData first, then perform any
 * modifications on headers based on merged cells in the sheet,
 * then convert to JSON.
 */
export async function convertXLSXToJSON<Data = GenericData, Output = object>({
  file,
  url,
  data,
  sheetIndex,
  sheetName,
  parsingOptions,
  preset,
  headers,
}: XLSXTableDataConfig): Promise<Output | Output[]> {
  const { sheet, tableData } = await convertXLSXToTableData<Data>({
    file,
    url,
    data,
    sheetIndex,
    sheetName,
    parsingOptions,
  });

  return convertTableDataToJSON<Data, Output>(tableData, {
    preset,
    headers,
    modifyHeaders: (headers) => {
      // Calculate header cell widths/heights for any
      // merged cells.
      if (sheet["!merges"] && sheet["!merges"].length) {
        headers.forEach((row) => {
          row.forEach((cell) => {
            // @ts-ignore
            sheet["!merges"].forEach((mergedCell) => {
              if (mergedCell.s.c === cell.c && mergedCell.s.r === cell.r) {
                cell.width = 1 + (mergedCell.e.c - mergedCell.s.c);
                cell.height = 1 + (mergedCell.e.r - mergedCell.s.r);
              }
            });
          });
        });
      }
    },
  });
}

export default convertXLSXToJSON;
