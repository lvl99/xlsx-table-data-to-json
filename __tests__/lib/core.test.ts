import fs from "node:fs";
import path from "node:path";
import xlsx from "xlsx";
import fetch from "node-fetch";
import { DEFAULT_XLSX_PARSING_OPTIONS } from "../../lib/constants";
import { convertXLSXToTableData, convertXLSXToJSON } from "../../lib/core";

jest.mock("node-fetch");
const { Response } = jest.requireActual("node-fetch");

const testXLSXFilePath = path.join(__dirname, "./test.xlsx");
const testXLSXFileBuffer = fs.readFileSync(testXLSXFilePath);
const testXLSXWorkBook = xlsx.readFile(
  testXLSXFilePath,
  DEFAULT_XLSX_PARSING_OPTIONS
);

const testCSVFilePath = path.join(__dirname, "./test.csv");
const testCSVFileBuffer = fs.readFileSync(testCSVFilePath);

describe("convertXlsxToTableData", () => {
  it("should open XLSX file (local path string) and convert contents to array table data", async () => {
    const { tableData } = await convertXLSXToTableData({
      file: testXLSXFilePath,
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(4);
    expect(tableData).toMatchSnapshot();
  });

  it("should open XLSX file (Buffer) and convert contents to array table data", async () => {
    const { tableData } = await convertXLSXToTableData({
      file: testXLSXFileBuffer,
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(4);
    expect(tableData).toMatchSnapshot();
  });

  it("should open CSV file (local path string) and convert contents to array table data", async () => {
    const { tableData } = await convertXLSXToTableData({
      file: testCSVFilePath,
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(4);
    expect(tableData).toMatchSnapshot();
  });

  it("should open CSV file (Buffer) and convert contents to array table data", async () => {
    const { tableData } = await convertXLSXToTableData({
      file: testCSVFileBuffer,
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(4);
    expect(tableData).toMatchSnapshot();
  });

  it("should open XLSX file (remote HTTP URL) and convert contents to array table data", async () => {
    // @ts-ignore
    fetch.mockReturnValue(Promise.resolve(new Response(testXLSXFileBuffer)));

    const { tableData } = await convertXLSXToTableData({
      url: "https://example.com/table-data.xlsx",
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(4);
    expect(tableData).toMatchSnapshot();
  });

  it("should use XLSX data (xlsx.WorkBook) and convert contents to array table data", async () => {
    const { tableData } = await convertXLSXToTableData({
      data: testXLSXWorkBook,
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(4);
    expect(tableData).toMatchSnapshot();
  });

  it("should throw an error if no file, url or data given", async () => {
    await expect(convertXLSXToTableData({})).rejects.toThrowError(
      "Please provide XLSX data, url or file (either file contents buffer or local path string)"
    );
  });

  it("should throw an error if something happened with fetch", async () => {
    const errorMsg = "Testing error in node-fetch";

    // @ts-ignore
    fetch.mockReturnValue(Promise.reject(new Error(errorMsg)));

    await expect(
      convertXLSXToTableData({ url: "https://example.com/test.xlsx" })
    ).rejects.toThrowError(errorMsg);
  });

  it("should throw an error if data is not xlsx.WorkBook", async () => {
    await expect(
      convertXLSXToTableData({ data: {} as xlsx.WorkBook })
    ).rejects.toThrowError(
      "Please provide XLSX data, url or file (either file contents buffer or local path string)"
    );
  });

  it("should pick out a specific sheet by zero-based index", async () => {
    const { tableData } = await convertXLSXToTableData({
      data: testXLSXWorkBook,
      sheetIndex: 1,
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(3);
    expect(tableData).toMatchSnapshot();
  });

  it("should pick out a specific sheet by name", async () => {
    const { tableData } = await convertXLSXToTableData({
      data: testXLSXWorkBook,
      sheetName: "column",
    });
    expect(tableData).toBeInstanceOf(Array);
    expect(tableData).toHaveLength(3);
    expect(tableData).toMatchSnapshot();
  });

  it("should convert special Excel cells to JSON data", async () => {
    const originalTableData = [
      [
        "test",
        "",
        undefined,
        null,
        true,
        false,
        123456789,
        new Date("2022-11-13"),
      ],
    ];
    const sheet = xlsx.utils.aoa_to_sheet(originalTableData, {
      cellDates: true,
    });
    const wb = xlsx.utils.book_new();
    xlsx.utils.book_append_sheet(wb, sheet, "Test 1");

    const { tableData } = await convertXLSXToTableData({
      data: wb,
    });
    expect(tableData).toHaveLength(1);
    expect(tableData[0]).toHaveLength(8);
    expect(tableData[0][0]).toBe("test");
    expect(tableData[0][1]).toBe("");
    expect(tableData[0][2]).toBeUndefined();
    expect(tableData[0][3]).toBeUndefined();
    expect(tableData[0][4]).toBe(true);
    expect(tableData[0][5]).toBe(false);
    expect(tableData[0][6]).toBe(123456789);
    expect(tableData[0][7]).toStrictEqual("2022-11-13T00:00:00.000Z");
  });
});

describe("convertXLSXToJSON", () => {
  it("should open XLSX file (local path string) and convert contents to array table data", async () => {
    const result = await convertXLSXToJSON({
      file: testXLSXFilePath,
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it("should open XLSX file (Buffer) and convert contents to array table data", async () => {
    const result = await convertXLSXToJSON({
      file: testXLSXFileBuffer,
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it("should open CSV file (local path string) and convert contents to array table data", async () => {
    const result = await convertXLSXToJSON({
      file: testCSVFilePath,
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it("should open CSV file (Buffer) and convert contents to array table data", async () => {
    const result = await convertXLSXToJSON({
      file: testCSVFileBuffer,
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it("should open XLSX file (remote HTTP URL) and convert contents to array table data", async () => {
    // @ts-ignore
    fetch.mockReturnValue(Promise.resolve(new Response(testXLSXFileBuffer)));

    const result = await convertXLSXToJSON({
      url: "https://example.com/table-data.xlsx",
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it("should use XLSX data (xlsx.Workbook) and convert contents to array table data", async () => {
    const result = await convertXLSXToJSON({
      data: testXLSXWorkBook,
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it("should throw an error if no file, url or data given", async () => {
    await expect(convertXLSXToJSON({})).rejects.toThrowError(
      "Please provide XLSX data, url or file (either file contents buffer or local path string)"
    );
  });

  it("should throw an error if something happened with fetch", async () => {
    const errorMsg = "Testing error in node-fetch";

    // @ts-ignore
    fetch.mockReturnValue(Promise.reject(new Error(errorMsg)));

    await expect(
      convertXLSXToJSON({ url: "https://example.com/test.xlsx" })
    ).rejects.toThrowError(errorMsg);
  });

  it("should throw an error if data is not xlsx.WorkBook", async () => {
    await expect(
      convertXLSXToJSON({ data: {} as xlsx.WorkBook })
    ).rejects.toThrowError(
      "Please provide XLSX data, url or file (either file contents buffer or local path string)"
    );
  });

  it("should pick out a specific sheet by zero-based index", async () => {
    const result = await convertXLSXToJSON({
      file: testXLSXFileBuffer,
      sheetIndex: 1,
      preset: "row.column",
    });
    expect(result).toBeInstanceOf(Object);
    expect(result).toMatchSnapshot();
  });

  it("should merge header cells", async () => {
    const result = await convertXLSXToJSON({
      file: testXLSXFileBuffer,
      sheetIndex: 4,
      preset: "row.row",
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it("should merge and nest header cells", async () => {
    const result = await convertXLSXToJSON({
      file: testXLSXFileBuffer,
      sheetIndex: 6,
      preset: "row.row",
    });
    expect(result).toBeInstanceOf(Array);
    expect(result).toHaveLength(2);
    expect(result).toMatchSnapshot();
  });
});
