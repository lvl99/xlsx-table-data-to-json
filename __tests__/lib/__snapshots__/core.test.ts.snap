// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertXLSXToJSON should merge and nest header cells 1`] = `
[
  {
    "Colour": "999B",
    "Ending": {
      "Amount": 150,
      "Units": 3,
    },
    "Reference": "AC027",
    "Shipped": {
      "Amount": 150,
      "Units": 3,
    },
    "Starting": {
      "Amount": 250,
      "Units": 5,
    },
    "Transferred": {
      "Amount": 50,
      "Units": 1,
    },
    "Unit Cost": 50,
  },
  {
    "Colour": "001W",
    "Ending": {
      "Amount": 120,
      "Units": 4,
    },
    "Reference": "AC027",
    "Shipped": {
      "Amount": 30,
      "Units": 1,
    },
    "Starting": {
      "Amount": 90,
      "Units": 3,
    },
    "Transferred": {
      "Amount": 60,
      "Units": 2,
    },
    "Unit Cost": 30,
  },
]
`;

exports[`convertXLSXToJSON should merge header cells 1`] = `
[
  {
    "Test A": {
      "Bob": 1,
      "John": 2,
      "Thom": 3,
    },
    "Test B": {
      "Bob": "a",
      "John": "b",
      "Thom": "c",
    },
    "Test C": {
      "Bob": "2021-12-30T23:00:00.000Z",
      "John": "2021-11-23T23:00:00.000Z",
      "Thom": "2021-10-11T22:00:00.000Z",
    },
  },
]
`;

exports[`convertXLSXToJSON should open CSV file (Buffer) and convert contents to array table data 1`] = `
[
  {
    "Test A": 1,
    "Test B": 2,
    "Test C": 3,
  },
  {
    "Test A": "a",
    "Test B": "b",
    "Test C": "c",
  },
  {
    "Test A": "2021-12-30T23:00:00.000Z",
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
]
`;

exports[`convertXLSXToJSON should open CSV file (local path string) and convert contents to array table data 1`] = `
[
  {
    "Test A": 1,
    "Test B": 2,
    "Test C": 3,
  },
  {
    "Test A": "a",
    "Test B": "b",
    "Test C": "c",
  },
  {
    "Test A": "2021-12-30T23:00:00.000Z",
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
]
`;

exports[`convertXLSXToJSON should open XLSX file (Buffer) and convert contents to array table data 1`] = `
[
  {
    "Test A": 1,
    "Test B": 2,
    "Test C": 3,
  },
  {
    "Test A": "a",
    "Test B": "b",
    "Test C": "c",
  },
  {
    "Test A": "2021-12-30T23:00:00.000Z",
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
]
`;

exports[`convertXLSXToJSON should open XLSX file (local path string) and convert contents to array table data 1`] = `
[
  {
    "Test A": 1,
    "Test B": 2,
    "Test C": 3,
  },
  {
    "Test A": "a",
    "Test B": "b",
    "Test C": "c",
  },
  {
    "Test A": "2021-12-30T23:00:00.000Z",
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
]
`;

exports[`convertXLSXToJSON should open XLSX file (remote HTTP URL) and convert contents to array table data 1`] = `
[
  {
    "Test A": 1,
    "Test B": 2,
    "Test C": 3,
  },
  {
    "Test A": "a",
    "Test B": "b",
    "Test C": "c",
  },
  {
    "Test A": "2021-12-30T23:00:00.000Z",
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
]
`;

exports[`convertXLSXToJSON should pick out a specific sheet by zero-based index 1`] = `
{
  "1": {
    "Test B": 2,
    "Test C": 3,
  },
  "2021-12-30T23:00:00.000Z": {
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
  "a": {
    "Test B": "b",
    "Test C": "c",
  },
}
`;

exports[`convertXLSXToJSON should use XLSX data (xlsx.Workbook) and convert contents to array table data 1`] = `
[
  {
    "Test A": 1,
    "Test B": 2,
    "Test C": 3,
  },
  {
    "Test A": "a",
    "Test B": "b",
    "Test C": "c",
  },
  {
    "Test A": "2021-12-30T23:00:00.000Z",
    "Test B": "2021-11-23T23:00:00.000Z",
    "Test C": "2021-10-11T22:00:00.000Z",
  },
]
`;

exports[`convertXlsxToTableData should open CSV file (Buffer) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should open CSV file (local path string) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should open XLSX file (Buffer) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
    ,
    ,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should open XLSX file (local path string) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
    ,
    ,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should open XLSX file (remote HTTP URL) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
    ,
    ,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should pick out a specific sheet by name 1`] = `
[
  [
    "Test A",
    1,
    "a",
    "2021-12-30T23:00:00.000Z",
  ],
  [
    "Test B",
    2,
    "b",
    "2021-11-23T23:00:00.000Z",
  ],
  [
    "Test C",
    3,
    "c",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should pick out a specific sheet by zero-based index 1`] = `
[
  [
    "Test A",
    1,
    "a",
    "2021-12-30T23:00:00.000Z",
  ],
  [
    "Test B",
    2,
    "b",
    "2021-11-23T23:00:00.000Z",
  ],
  [
    "Test C",
    3,
    "c",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should use XLSX data (xlsx.WorkBook) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
    ,
    ,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;

exports[`convertXlsxToTableData should use XLSX data (xlsx.Workbook) and convert contents to array table data 1`] = `
[
  [
    "Test A",
    "Test B",
    "Test C",
    ,
    ,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  [
    1,
    2,
    3,
  ],
  [
    "a",
    "b",
    "c",
  ],
  [
    "2021-12-30T23:00:00.000Z",
    "2021-11-23T23:00:00.000Z",
    "2021-10-11T22:00:00.000Z",
  ],
]
`;
